### PR TITLE
Add promo code groups

### PR DESCRIPTION
**Pull request by kitsuta**
Fixes https://jira.magfest.net/browse/MAGDEV-421.

### CODE REVIEW

File: uber/templates/regform.html [hunk 1]
```diff
@@ -614,7 +614,7 @@ <h4>{{ tier.name }}</h4>
 <div class="form-group">
   <div class="checkbox col-sm-9 col-sm-offset-3">
     <label for="pii_consent" class="checkbox-label">
-    <input type="checkbox" name="pii_consent" id="pii_consent" value="1" required />
+    <input type="checkbox" name="pii_consent" id="pii_consent" value="1" required {% if pii_consent %}checked{% endif %} />
```
Comment (kitsuta) [inline uber/templates/regform.html]: Unrelated QoL change that keeps the pii_consent checkbox checked during prereg.

File: uber/utils.py [hunk 2]
```diff
@@ -825,24 +825,27 @@ def get_unpaid_promo_code_uses_count(cls, id, already_counted_attendee_ids=None)
...
             return [cls.to_sessionized(t) for t in m]
         elif isinstance(m, dict):
             return m
         elif isinstance(m, Attendee):
-            return m.to_dict(
+            d = m.to_dict(
                 Attendee.to_dict_default_attrs
                 + ['promo_code']
                 + list(Attendee._extra_apply_attrs_restricted))
+            d['name'] = name
```
Comment (kitsuta) [inline uber/utils.py]: This implementation is kind of janky due to the nature of promo code groups -- as far as money goes, the Attendee object needs to be the one who pays for a promo code group, but we still need to sessionize and desessionize certain group-specific info. We're doing that here by hijacking the Attendee dict, which can hold extra variables.

The real fix for this is to implement a cart-like system and make PromoCodeGroups a piece of merch. Sadly, I did not have time to do this.

File: uber/utils.py [hunk 3]
```diff
@@ -825,24 +825,27 @@ def get_unpaid_promo_code_uses_count(cls, id, already_counted_attendee_ids=None)
...
+            d['badges'] = badges
+            return d
         elif isinstance(m, Group):
             return m.to_dict(
                 Group.to_dict_default_attrs
                 + ['attendees']
                 + list(Group._extra_apply_attrs_restricted))
         else:
-            raise AssertionError('{} is not an attendee or group'.format(m))
+            raise AssertionError('{} is not an attendee or dict'.format(m))
```
Comment (kitsuta) [inline uber/utils.py]: Whoops, this needs to be reverted -- there was a brief, heady period of time where I had removed all the Group object stuff from the Charge class... right until I tried to load a group_members page and found we still needed it. :'D

File: uber/utils.py [hunk 4]
```diff
@@ -913,9 +927,17 @@ def description(self):
     def receipt_email(self):
         return self.models[0].email if self.models and self.models[0].email else self._receipt_email
 
-    @cached_property
+    @property
```
Comment (kitsuta) [inline uber/utils.py]: Also need to revert this.

File: uber/models/__init__.py [hunk 5]
```diff
@@ -846,9 +870,25 @@ def lookup_promo_code(self, code):
             except Exception:
                 pass
             else:
-                clause = clause.or_(PromoCode.id == promo_code_id)
+                clause = clause.or_(model.id == promo_code_id)
+
+            return self.query(model).filter(clause).order_by(model.normalized_code.desc()).first()
+
+        def create_promo_code_group(self, attendee, name, badges):
```
[the remark this answers is not in the document]
Comment (kitsuta) [reply uber/models/__init__.py]: I agree, I guess my issue with doing this is that we end up putting a lot of class-specific functions in the Session class, which is A) kinda hard to find if you don't know we're doing that and 2) going to clog up the Session class with a lot of sorta-random stuff. Like, logically speaking these functions don't deal with the session so much as they deal with PromoCode and PromoCodeGroup's relationship _to_ the session. I don't know how we'd organize the code to reflect that.

[the remark this answers is not in the document]
Comment (EliAndrewC) [reply uber/models/__init__.py]: Yeah, that's a fair point.  If we wanted to, we could add another decorator to transplant these things into ``SessionMixin``, e.g. in ``attendee.py`` we could have something like

```
@session_monkeypatch
class SessionMixin:
    def create_promo_code_group(self, attendee, name, badges):
        ...
```

so that instead of having one giant ``SessionMixin`` class we instead have a bunch of local ones located in whichever files have the associated models.

Definitely out of scope for this ticket, but something to keep in mind if you ever feel like ``SessionMixin`` needs to be broken up.

**Comment by kitsuta** [comment]
Our unit tests are so broken that @EliAndrewC and I have spent multiple hours just trying to get them in a working state... and with another new issue cropping up tonight once I started digging into them, we're going to call it and delay the unit test portion of this PR until we can get them properly fixed (right now we can _sort of_ get them to run). Merging now.
